pkg/transformers/openshift/openshift.go

```diff
@@ -0,0 +1,655 @@
+	pipelines    app.PipelineGroup
+	containers   map[string]*kapi.Container
+	objects      app.Objects
+	bases        []string
+	aliases      map[string]sets.String
+}
+
+func Transform(p *project.Project, bases []string) error {
+	data := TransformData{
+		errs:         []error{},
+		warnings:     make(map[string][]string),
+		serviceOrder: sets.NewString(),
+		joins:        make(map[string]sets.String),
+		volumesFrom:  make(map[string][]string),
+		colocated:    []sets.String{},
+		builds:       make(map[string]*app.Pipeline),
+		pipelines:    app.PipelineGroup{},
+		containers:   make(map[string]*kapi.Container),
+		objects:      app.Objects{},
+		bases:        bases,
+		aliases:      make(map[string]sets.String),
+	}
+
+	updateServiceOrder(p, &data)
+	updateVolumes(p, &data)
+
+	err := updatePodsList(p, &data)
+	if err != nil {
+		return err
+	}
+
+	updateAliases(p, &data)
+
+	g := app.NewImageRefGenerator()
+
+	err = updateBuildPipeline(p, g, &data)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("After upateBuildPipeline")
+	fmt.Println("pipelines: ", data.pipelines)
+	fmt.Println("builds: ", data.builds)
+
+	err = updateDeploymentConfigs(p, g, &data)
+	if err != nil {
+		return err
+	}
+
+	if len(data.errs) > 0 {
+		return utilerrs.NewAggregate(data.errs)
+	}
+
+	err = updatePipelineObjects(&data)
+	if err != nil {
+		return err
+	}
+
+	err = updateServiceObjects(&data)
+
+	template := &templateapi.Template{}
+	template.Name = p.Name
+
+	// for each container that defines VolumesFrom, copy equivalent mounts.
+	// TODO: ensure mount names are unique?
+	for target, otherContainers := range data.volumesFrom {
+		for _, from := range otherContainers {
+			for _, volume := range data.containers[from].VolumeMounts {
+				data.containers[target].VolumeMounts = append(data.containers[target].VolumeMounts, volume)
+			}
+		}
+	}
+
+	template.Objects = data.objects
+
+	// generate warnings
+	if len(data.warnings) > 0 {
+		allWarnings := sets.NewString()
+		for msg, services := range data.warnings {
+			allWarnings.Insert(fmt.Sprintf("%s: %s", strings.Join(services, ","), msg))
+		}
+		if template.Annotations == nil {
+			template.Annotations = make(map[string]string)
+		}
+		template.Annotations[app.GenerationWarningAnnotation] = fmt.Sprintf("not all docker-compose fields were honored:\n* %s", strings.Join(allWarnings.List(), "\n* "))
+	}
+
+	var convErr error
+	template.Objects, convErr = convertToVersion(template.Objects, "v1")
+	if convErr != nil {
+		panic(convErr)
+	}
+
+	// make it List instead of Template
+	list := &kapi.List{Items: template.Objects}
+
+	printer, _, _err := kubectl.GetPrinter("yaml", "")
+	if _err != nil {
+		panic(_err)
+	}
+	version := unversioned.GroupVersion{Group: "", Version: "v1"}
+	printer = kubectl.NewVersionedPrinter(printer, kapi.Scheme, version)
+	printer.PrintObj(list, os.Stdout)
+
+	return nil
+}
+
+// Get ordered services
+func updateServiceOrder(p *project.Project, data *TransformData) {
```

Comment:
most of function in this file are operating on `TransformData` struct. It would be nice to attach those functions to `TransfromData` stuct.
```go
func (data *TransformData)  updateServiceOrder(p *project.Project)
```